Low: alerts: send all MIB OIDs with all SNMP alerts

kgaillot opened the pull request:
Some SNMP trap handlers expect all OID fields from a MIB to be available with every trap.

kgaillot commented:
@HideoYamauchi , 

Would you mind reviewing this change for me? It was suggested by a RHEL user who has an SNMP trap sink that requires that every trap supply all OIDs in its MIB. Since you have better SNMP knowledge than I do, I thought you would be the best person to review it. Do you think it is a good idea?

HideoYamauchi commented:
Hi Ken,

I will review your fix and try checking the actual sending.
Can you get the time to confirm until the beginning of next week?

Best Regards,
Hideo Yamauchi.

HideoYamauchi commented:
Hi Ken,

I tried to look a bit behaviors, but it seems to be necessary to modify only one point.
As for the correction, I will report after confirming the operation in detail.

Best Regards,
Hideo Yamauchi.

kgaillot commented:
Thanks! Yes, there's no rush, it can wait.

HideoYamauchi commented:
Okay!
All right!

HideoYamauchi commented on extra/alerts/alert_snmp.sh.sample:
```diff
@@ -116,6 +93,24 @@ is_in_list() {
+        "PACEMAKER-MIB::pacemakerNotificationTargetReturnCode" i "$7" \
+        "PACEMAKER-MIB::pacemakerNotificationAttributeName" s "$8" \
+        "PACEMAKER-MIB::pacemakerNotificationAttributeValue" s "$9" \
+        "${hires_timestamp}" 2>&1
```
This double quotation seems unnecessary.
If it exists, traps can not be sent.

```
${hires_timestamp} 2>&1
```

kgaillot replied:
Thank you! I should have caught that. :-/ I've updated the request with this fix.

kgaillot commented:
I was unsure what values to use for "not applicable", too.

The MIB uses the SNMP Integer32 type, so maybe we could use the minimum or maximum value, i.e. -2147483648 or 2147483647.

For the string properties, maybe "n/a" (for "not applicable") would be appropriate. The "/" would not be legal in a node, resource, operation, or attribute name, so it could never conflict with a real one.

HideoYamauchi commented:
Hi Ken,

Okay!

I will set the value to be ignored to 2147483647 and "n / a" and confirm trap transmission.

Best Regards,
Hideo Yamauchi.

HideoYamauchi commented:
Hi Ken,

I changed and confirmed it as follows.

```
(snip)
case "$CRM_alert_kind" in
    node)
(snip)
        # <node> <rsc> <op> <desc> <status> <rc> <target-rc> <attr-name> <attr-value>
        output=$(send_pacemaker_trap "${CRM_alert_node}" "n/a" "n/a" \
            "${CRM_alert_desc}" 2147483647 2147483647 2147483647 "n/a" "n/a")
        rc=$?
        ;;
    fencing)
(snip)
        # <node> <rsc> <op> <desc> <status> <rc> <target-rc> <attr-name> <attr-value>
        output=$(send_pacemaker_trap "${CRM_alert_node}" "n/a" \
            "${CRM_alert_task}" "${CRM_alert_desc}" 2147483647 \
            ${CRM_alert_rc} 2147483647 "n/a" "n/a")
        rc=$?
        ;;
    resource)
(snip)
            # <node> <rsc> <op> <desc> <status> <rc> <target-rc> <attr-name> <attr-value>
            output=$(send_pacemaker_trap "${CRM_alert_node}" \
                "${CRM_alert_rsc}" "${CRM_alert_task}" "${CRM_alert_desc}" \
                ${CRM_alert_status} ${CRM_alert_rc} ${CRM_alert_target_rc} \
                "n/a" "n/a")
            rc=$?
            ;;
        esac
        ;;
    attribute)
        # <node> <rsc> <op> <desc> <status> <rc> <target-rc> <attr-name> <attr-value>
        output=$(send_pacemaker_trap "${CRM_alert_node}" "n/a" "n/a" "n/a" \
            2147483647 2147483647 2147483647 "${CRM_alert_attribute_name}" \
            "${CRM_alert_attribute_value}")
        rc=$?
        ;;
(snip)
```

A trap was sent as follows.
No problem.

 * trap_add_hires_timestamp_oid="true"
```
--nodes--
Sep 19 08:25:53 snmp-server snmptrapd[6470]: 2018-09-19 08:25:53 <UNKNOWN> [UDP: [192.168.40.110]:60931->[192.168.40.9]:162]:#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (109645201) 12 days, 16:34:12.01#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "lost"#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"#011HOST-RESOURCES-MIB::hrSystemDate = STRING: 2018-9-19,8:25:53.7

--fencing--
Sep 19 08:25:58 snmp-server snmptrapd[6470]: 2018-09-19 08:25:58 <UNKNOWN> [UDP: [192.168.40.110]:42276->[192.168.40.9]:162]:#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (109645723) 12 days, 16:34:17.23#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "st_notify_fence"#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "Operation reboot of rh74-01 by rh74-02 for pacemaker-controld.17694@rh74-02: OK (ref=51439bc7-b204-4373-871f-a8192a52a6ff)"#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 0#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"#011HOST-RESOURCES-MIB::hrSystemDate = STRING: 2018-9-19,8:25:59.0
--resources--
2018-09-19 08:25:07 <UNKNOWN> [UDP: [192.168.40.100]:38382->[192.168.40.9]:162]:#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (123882) 0:20:38.82#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "prmDummy"#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "start"#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "ok"#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 0#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 0#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 0#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"#011HOST-RESOURCES-MIB::hrSystemDate = STRING: 2018-9-19,8:25:7.6

--attributes--
Sep 19 08:27:21 snmp-server snmptrapd[6537]: 2018-09-19 08:27:21 <UNKNOWN> [UDP: [192.168.40.100]:46062->[192.168.40.9]:162]:#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (7988) 0:01:19.88#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "test"#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "100"#011HOST-RESOURCES-MIB::hrSystemDate = STRING: 2018-9-19,8:27:21.7

```

 * trap_add_hires_timestamp_oid="false" 
```
--nodes--
Sep 19 08:04:20 snmp-server snmptrapd[6357]: 2018-09-19 08:04:20 <UNKNOWN> [UDP: [192.168.40.110]:41397->[192.168.40.9]:162]:#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (109515894) 12 days, 16:12:38.94#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "lost"#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"

--fencing--
Sep 19 08:04:25 snmp-server snmptrapd[6357]: 2018-09-19 08:04:25 <UNKNOWN> [UDP: [192.168.40.110]:46653->[192.168.40.9]:162]:#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (109516418) 12 days, 16:12:44.18#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "st_notify_fence"#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "Operation reboot of rh74-01 by rh74-02 for pacemaker-controld.16676@rh74-02: OK (ref=f078e623-40f4-4308-96ed-50a20b7365f4)"#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 0#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"

--resources--
2018-09-19 08:02:58 <UNKNOWN> [UDP: [192.168.40.100]:52623->[192.168.40.9]:162]:#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (8559182) 23:46:31.82#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "prmDummy"#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "start"#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "ok"#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 0#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 0#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 0#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"

--attributes--
2018-09-19 08:17:52 <UNKNOWN> [UDP: [192.168.40.100]:50222->[192.168.40.9]:162]:#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (80404) 0:13:24.04#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "n/a"#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "test"#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "100"
```

Best Regards,
Hideo Yamauchi.

HideoYamauchi commented:
Hi Ken,

Ultimately I confirmed the trap with the following change.
I decided that it is not necessary to use it as a parameter.

```
(snip)
trap_snmp_persistent_dir_default="/var/lib/pacemaker/snmp"
trap_ignore_int32_field=2147483647
tarp_ignore_string_field="n/a"
(snip)
case "$CRM_alert_kind" in
    node)
        is_in_list "${trap_node_states}" "${CRM_alert_desc}"
        [ $? -ne 0 ] && exit 0

        # <node> <rsc> <op> <desc> <status> <rc> <target-rc> <attr-name> <attr-value>
        output=$(send_pacemaker_trap "${CRM_alert_node}" ${tarp_ignore_string_field} ${tarp_ignore_string_field} \
            "${CRM_alert_desc}" ${trap_ignore_int32_field} ${trap_ignore_int32_field} ${trap_ignore_int32_field} ${tarp_ignore_string_field} ${tarp_ignore_string_field})
        rc=$?
        ;;
    fencing)
        is_in_list "${trap_fencing_tasks}" "${CRM_alert_task}"
        [ $? -ne 0 ] && exit 0

        # <node> <rsc> <op> <desc> <status> <rc> <target-rc> <attr-name> <attr-value>
        output=$(send_pacemaker_trap "${CRM_alert_node}" ${tarp_ignore_string_field} \
            "${CRM_alert_task}" "${CRM_alert_desc}" ${trap_ignore_int32_field} \
            ${CRM_alert_rc} ${trap_ignore_int32_field} ${tarp_ignore_string_field} ${tarp_ignore_string_field})
        rc=$?
        ;;
    resource)
        is_in_list "${trap_resource_tasks}" "${CRM_alert_task}"
        [ $? -ne 0 ] && exit 0

        case "${CRM_alert_desc}" in
        Cancelled) ;;
        *)
            if [ "${trap_monitor_success}" = "false" ] \
                && [ "${CRM_alert_rc}" = "${CRM_alert_target_rc}" ] \
                && [ "${CRM_alert_task}" = "monitor" ]; then
                    exit
            fi

            # <node> <rsc> <op> <desc> <status> <rc> <target-rc> <attr-name> <attr-value>
            output=$(send_pacemaker_trap "${CRM_alert_node}" \
                "${CRM_alert_rsc}" "${CRM_alert_task}" "${CRM_alert_desc}" \
                ${CRM_alert_status} ${CRM_alert_rc} ${CRM_alert_target_rc} \
                ${tarp_ignore_string_field} ${tarp_ignore_string_field})
            rc=$?
            ;;
        esac
        ;;
    attribute)
        # <node> <rsc> <op> <desc> <status> <rc> <target-rc> <attr-name> <attr-value>
        output=$(send_pacemaker_trap "${CRM_alert_node}" ${tarp_ignore_string_field} ${tarp_ignore_string_field} ${tarp_ignore_string_field} \
            ${trap_ignore_int32_field} ${trap_ignore_int32_field} ${trap_ignore_int32_field} "${CRM_alert_attribute_name}" \
            "${CRM_alert_attribute_value}")
        rc=$?
        ;;
(snip)
```

Best Regards,
Hideo Yamauchi.

kgaillot commented:
@HideoYamauchi ,

Thanks for all the comments and testing! I have updated the request to include the suggested default values. I also simplified the new code and made the values user-configurable.

HideoYamauchi commented on extra/alerts/alert_snmp.sh.sample:
```diff
@@ -124,77 +145,36 @@ if [ -z ${SNMP_PERSISTENT_DIR} ]; then
-            fi
+        if is_in_list "${trap_resource_tasks}" "${CRM_alert_task}" && \
+           [ "${CRM_alert_desc}" -ne "Cancelled" ] ; then
 
```
We need to fix this.
```
           [ "x${CRM_alert_desc}" != "xCancelled" ] ; then
```

HideoYamauchi commented:
Hi Ken,

One place correction was necessary.

```
           [ "x${CRM_alert_desc}" != "xCancelled" ] ; then

```

I confirmed that a trap will be sent as follows.


 1. trap_add_hires_timestamp_oid="true"
```
--nodes--
Sep 20 08:59:46 snmp-server snmptrapd[7506]: 2018-09-20 08:59:46 <UNKNOWN> [UDP: [192.168.40.110]:43915->[192.168.40.9]:162]:
#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (118488568) 13 days, 17:08:05.68
#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap
#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"
#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "lost"
#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"
#011HOST-RESOURCES-MIB::hrSystemDate = STRING: 2018-9-20,8:59:47.4

--fencing--
Sep 20 08:59:52 snmp-server snmptrapd[7506]: 2018-09-20 08:59:52 <UNKNOWN> [UDP: [192.168.40.110]:59470->[192.168.40.9]:162]:
#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (118489095) 13 days, 17:08:10.95
#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap
#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"
#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "st_notify_fence"
#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "Operation reboot of rh74-01 by rh74-02 for pacemaker-controld.2201@rh74-02: OK (ref=7c03e365-2b8b-4e0c-8579-ee4600c6cbf9)"
#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 0
#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"
#011HOST-RESOURCES-MIB::hrSystemDate = STRING: 2018-9-20,8:59:52.7

--resources--
Sep 20 08:58:09 snmp-server snmptrapd[7471]: 2018-09-20 08:58:09 <UNKNOWN> [UDP: [192.168.40.100]:44646->[192.168.40.9]:162]:
#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (8832797) 1 day, 0:32:07.97
#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap
#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"
#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "prmDummy"
#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "start"
#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "ok"
#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 0
#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 0
#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 0
#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"
#011HOST-RESOURCES-MIB::hrSystemDate = STRING: 2018-9-20,8:58:9.8

--attributes--
Sep 20 09:03:22 snmp-server snmptrapd[7559]: 2018-09-20 09:03:22 <UNKNOWN> [UDP: [192.168.40.100]:46652->[192.168.40.9]:162]:
#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (20762) 0:03:27.62
#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap
#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"
#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "test"
#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "100"
#011HOST-RESOURCES-MIB::hrSystemDate = STRING: 2018-9-20,9:3:22.4

```

 2. trap_add_hires_timestamp_oid="false"
```
--nodes--
Sep 20 09:05:53 snmp-server snmptrapd[7594]: 2018-09-20 09:05:53 <UNKNOWN> [UDP: [192.168.40.110]:45157->[192.168.40.9]:162]:
#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (118525231) 13 days, 17:14:12.31
#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap
#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"
#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "lost"
#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"

--fencing--
Sep 20 09:05:58 snmp-server snmptrapd[7594]: 2018-09-20 09:05:58 <UNKNOWN> [UDP: [192.168.40.110]:59328->[192.168.40.9]:162]:
#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (118525753) 13 days, 17:14:17.53
#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap
#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"
#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "st_notify_fence"
#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "Operation reboot of rh74-01 by rh74-02 for pacemaker-controld.3409@rh74-02: OK (ref=cf4229b1-231c-4625-9758-0f5c25bdcf17)"
#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 0
#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"

--resources--
Sep 20 09:04:50 snmp-server snmptrapd[7594]: 2018-09-20 09:04:50 <UNKNOWN> [UDP: [192.168.40.100]:58743->[192.168.40.9]:162]:
#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (29566) 0:04:55.66
#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap
#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"
#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "prmDummy"
#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "start"
#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "ok"
#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 0
#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 0
#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 0
#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "n/a"

--attributes--
Sep 20 09:07:58 snmp-server snmptrapd[7629]: 2018-09-20 09:07:58 <UNKNOWN> [UDP: [192.168.40.100]:53709->[192.168.40.9]:162]:
#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (11641) 0:01:56.41
#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap
#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-01"
#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "n/a"
#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: 2147483647
#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "test"
#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "100"
```


Also, I confirmed the transmission by also specifying the trap_ignore_int32 and trap_ignore_string parameters. 
We only confirmed the traps of nodes.(trap_add_hires_timestamp_oid="true")

```
--attributes setting--
(snip)
alert notify_9 /usr/share/pacemaker/alerts/alert_snmp.sh \
        meta \
        kind="node,fencing,resource,attribute" \
        attributes \
        trap_resource_tasks="start,stop,monitor,promote,demote" \
        trap_add_hires_timestamp_oid="true" \
        trap_ignore_int32=-2147483648 \
        trap_ignore_string="not applicable" \
        to "192.168.40.9"
(snip)

--nodes--
Sep 20 08:40:47 snmp-server snmptrapd[7433]: 2018-09-20 08:40:47 <UNKNOWN> [UDP: [192.168.40.100]:48514->[192.168.40.9]:162]:
#012DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (8728630) 1 day, 0:14:46.30
#011SNMPv2-MIB::snmpTrapOID.0 = OID: PACEMAKER-MIB::pacemakerNotificationTrap
#011PACEMAKER-MIB::pacemakerNotificationNode = STRING: "rh74-02"
#011PACEMAKER-MIB::pacemakerNotificationResource = STRING: "not applicable"
#011PACEMAKER-MIB::pacemakerNotificationOperation = STRING: "not applicable"
#011PACEMAKER-MIB::pacemakerNotificationDescription = STRING: "lost"
#011PACEMAKER-MIB::pacemakerNotificationStatus = INTEGER: -2147483648
#011PACEMAKER-MIB::pacemakerNotificationReturnCode = INTEGER: -2147483648
#011PACEMAKER-MIB::pacemakerNotificationTargetReturnCode = INTEGER: -2147483648
#011PACEMAKER-MIB::pacemakerNotificationAttributeName = STRING: "not applicable"
#011PACEMAKER-MIB::pacemakerNotificationAttributeValue = STRING: "not applicable"
#011HOST-RESOURCES-MIB::hrSystemDate = STRING: 2018-9-20,8:40:48.1

```

There was one point fix, but all the traps were correctly transmitted.

Best Regards,
Hideo Yamauchi.

kgaillot commented:
@HideoYamauchi , thanks for all your help!

HideoYamauchi commented:
Hi Ken,

No problem.
Thanks!

Hideo Yamauchi.

jnpkrn commented:
Btw. do we know which trap consumer exhibits this picky behaviour
(at least compared to standard facilities like snmpd otherwise it
would have been spotted earlier?), so passing along that would be
a good verification step in the future when easily obtainable?

kgaillot commented:
no, I think it may have been a custom or proprietary trap handler at a user's site
